src/content/docs/guides/data-shaping/deduplicate-events.mdx

```diff
@@ -3,73 +3,178 @@ title: Deduplicate events
+from {id: {orig_h: "10.0.0.1", resp_h: "192.168.1.1"}, bytes: 1024},
+     {id: {orig_h: "10.0.0.2", resp_h: "192.168.1.1"}, bytes: 2048},
+     {id: {orig_h: "10.0.0.1", resp_h: "192.168.1.1"}, bytes: 512},
+     {id: {orig_h: "10.0.0.1", resp_h: "192.168.1.2"}, bytes: 256}
 deduplicate {orig_h: id.orig_h, resp_h: id.resp_h}
 ```
 
-Providing `id.orig_h` and `id.resp_h` to the operator restricts the output to
-all unique host pairs. Note that flipped connections occur twice here, i.e., A →
-B as well as B → A are present.
+```tql
+{id: {orig_h: "10.0.0.1", resp_h: "192.168.1.1"}, bytes: 1024}
+{id: {orig_h: "10.0.0.2", resp_h: "192.168.1.1"}, bytes: 2048}
+{id: {orig_h: "10.0.0.1", resp_h: "192.168.1.2"}, bytes: 256}
+```
 
-## Remove duplicate alerts
+The deduplication works on the extracted host pairs:
 
-Are you're overloaded with alerts, like every analyst? Let's remove some noise
-from our alerts.
+- First connection (10.0.0.1 → 192.168.1.1) is kept
+- Second connection (10.0.0.2 → 192.168.1.1) is kept (different origin)
+- Third connection (10.0.0.1 → 192.168.1.1) is dropped (duplicate of first)
+- Fourth connection (10.0.0.1 → 192.168.1.2) is kept (different destination)
+
+Note that flipped connections (A→B vs B→A) are considered different pairs. This
+helps identify bidirectional communication patterns.
+
+## Remove duplicate alerts
 
-First, let's check what our alert dataset looks like:
+Security monitoring often generates duplicate alerts that create noise and fatigue.
+Here's how to suppress repeated alerts for the same threat pattern:
 
 ```tql
-export
-where @schema == "suricata.alert"
-top alert.signature
-head 5
+from {src_ip: "10.0.0.1", dest_ip: "8.8.8.8", signature: "Suspicious DNS", time: 1},
+     {src_ip: "10.0.0.1", dest_ip: "8.8.8.8", signature: "Suspicious DNS", time: 2},
+     {src_ip: "10.0.0.2", dest_ip: "8.8.8.8", signature: "Suspicious DNS", time: 3},
+     {src_ip: "10.0.0.1", dest_ip: "8.8.8.8", signature: "Port Scan", time: 4}
+deduplicate {src: src_ip, dst: dest_ip, sig: signature}
 ```
 
 ```tql
-{
-  alert.signature: "ET MALWARE Cobalt Strike Beacon Observed",
-  count: 117369,
-}
-{
-  alert.signature: "SURICATA STREAM ESTABLISHED packet out of window",
-  count: 103198,
-}
-{
-  alert.signature: "SURICATA STREAM Packet with invalid ack",
-  count: 21960,
-}
-{
-  alert.signature: "SURICATA STREAM ESTABLISHED invalid ack",
-  count: 21920,
-}
-{
-  alert.signature: "ET JA3 Hash - [Abuse.ch] Possible Dridex",
-  count: 16870,
-}
+{src_ip: "10.0.0.1", dest_ip: "8.8.8.8", signature: "Suspicious DNS", time: 1}
+{src_ip: "10.0.0.2", dest_ip: "8.8.8.8", signature: "Suspicious DNS", time: 3}
+{src_ip: "10.0.0.1", dest_ip: "8.8.8.8", signature: "Port Scan", time: 4}
 ```
 
-Hundreds of thousands of alerts! Maybe I'm just interested in one per hour per
-host affected host pair? Here's the pipeline for this:
+The deduplication creates a composite key from source, destination, and signature:
```

Comment:
What is a "composite key"? Isn't this just a more concise method to rename the fields?